File: go/orchestration.go
package aegis

import (
	"context"
	"fmt"

	"github.com/ai-aegis-yatav/aegis-sdk/go/internal"
)

// OrchestrationService wraps /v1/orchestration/* endpoints.
// Every runs/* call always includes v1/judge.
type OrchestrationService struct {
	t *internal.Transport
}

type OrchestrationRunRequest struct {
	Content   string             `json:"content"`
	Scenario  string             `json:"scenario,omitempty"`
	Mode      string             `json:"mode,omitempty"`
	Algorithm string             `json:"algorithm,omitempty"`
	Weights   map[string]float64 `json:"weights,omitempty"`
	Threshold *float64           `json:"threshold,omitempty"`
}

type OrchestrationContributor struct {
	Name      string  `json:"name"`
	Score     float64 `json:"score"`
	Block     bool    `json:"block"`
	LatencyMs *uint64 `json:"latency_ms,omitempty"`
}

type OrchestratedDecision struct {
	Scenario       string                     `json:"scenario"`
	Mode           string                     `json:"mode"`
	Algorithm      string                     `json:"algorithm"`
	Decision       string                     `json:"decision"`
	EnsembleScore  float64                    `json:"ensemble_score"`
	Threshold      float64                    `json:"threshold"`
	Contributors   []OrchestrationContributor `json:"contributors"`
	Explanation    string                     `json:"explanation"`
	TotalLatencyMs uint64                     `json:"total_latency_ms"`
}

type orchestrationRunResponse struct {
	Decision OrchestratedDecision `json:"decision"`
}

func (s *OrchestrationService) run(ctx context.Context, path string, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	var resp orchestrationRunResponse
	if err := s.t.Do(ctx, "POST", path, req, &resp); err != nil {
		return nil, err
	}
	return &resp.Decision, nil
}

func (s *OrchestrationService) Run(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs", req)
}
func (s *OrchestrationService) Basic(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/basic", req)
}
func (s *OrchestrationService) Standard(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/standard", req)
}
func (s *OrchestrationService) Full(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/full", req)
}
func (s *OrchestrationService) Advanced(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/advanced", req)
}
func (s *OrchestrationService) Agent(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/agent", req)
}
func (s *OrchestrationService) Anomaly(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/anomaly", req)
}
func (s *OrchestrationService) PII(ctx context.Context, req OrchestrationRunRequest) (*OrchestratedDecision, error) {
	return s.run(ctx, "/v1/orchestration/runs/pii", req)
}

type TimeseriesAnomalyRequest struct {
	Value     float64   `json:"value"`
	History   []float64 `json:"history"`
	Algorithm string    `json:"algorithm,omitempty"`
}

type TimeseriesAnomalyResponse struct {
	IsAnomaly    bool    `json:"is_anomaly"`
	AnomalyScore float64 `json:"anomaly_score"`
	Algorithm    string  `json:"algorithm"`
	LatencyMs    uint64  `json:"latency_ms"`
}

func (s *OrchestrationService) AnomalyTimeseries(ctx context.Context, req TimeseriesAnomalyRequest) (*TimeseriesAnomalyResponse, error) {
	var resp TimeseriesAnomalyResponse
	if err := s.t.Do(ctx, "POST", "/v1/orchestration/runs/anomaly/timeseries", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

type OrchestrationConfig struct {
	TenantID    string             `json:"tenant_id"`
	Scenario    string             `json:"scenario"`
	Algorithm   string             `json:"algorithm"`
	Mode        string             `json:"mode"`
	Weights     map[string]float64 `json:"weights"`
	Thresholds  map[string]float64 `json:"thresholds"`
	SourceJobID *string            `json:"source_job_id,omitempty"`
	UpdatedAt   string             `json:"updated_at"`
}

type UpsertConfigRequest struct {
	Algorithm   string             `json:"algorithm"`
	Mode        string             `json:"mode"`
	Weights     map[string]float64 `json:"weights,omitempty"`
	Thresholds  map[string]float64 `json:"thresholds,omitempty"`
	SourceJobID *string            `json:"source_job_id,omitempty"`
}

func (s *OrchestrationService) GetConfig(ctx context.Context, tenantID, scenario string) (*OrchestrationConfig, error) {
	var resp OrchestrationConfig
	if err := s.t.Do(ctx, "GET", fmt.Sprintf("/v1/orchestration/configs/%s/%s", tenantID, scenario), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OrchestrationService) UpsertConfig(ctx context.Context, tenantID, scenario string, body UpsertConfigRequest) (*OrchestrationConfig, error) {
	var resp OrchestrationConfig
	if err := s.t.Do(ctx, "PUT", fmt.Sprintf("/v1/orchestration/configs/%s/%s", tenantID, scenario), body, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GridSearchRequest is the body for the synchronous /v1/orchestration/gridsearch endpoint.
type GridSearchRequest struct {
	Domain     string `json:"domain"`
	Scenario   string `json:"scenario"`
	MaxSamples *int   `json:"max_samples,omitempty"`
}

type GridSearchEntry struct {
	Label     string  `json:"label"`
	Algorithm string  `json:"algorithm"`
	Precision float64 `json:"precision"`
	Recall    float64 `json:"recall"`
	F1        float64 `json:"f1"`
	PrAuc     float64 `json:"pr_auc"`
}

type GridSearchResponse struct {
	Domain       string            `json:"domain"`
	TotalSamples int               `json:"total_samples"`
	TopCombos    []GridSearchEntry `json:"top_combos"`
}

// Gridsearch runs a synchronous grid search over a domain dataset.
// Use CreateGridsearchJob for the async, DB-persisted variant.
func (s *OrchestrationService) Gridsearch(ctx context.Context, req GridSearchRequest) (*GridSearchResponse, error) {
	var resp GridSearchResponse
	if err := s.t.Do(ctx, "POST", "/v1/orchestration/gridsearch", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

type GridSearchJobRequest struct {
	Domain     string `json:"domain"`
	Scenario   string `json:"scenario"`
	MaxSamples *int   `json:"max_samples,omitempty"`
}

type CreateJobResponse struct {
	JobID  string `json:"job_id"`
	Status string `json:"status"`
}

type GridSearchJob struct {
	ID          string  `json:"id"`
	TenantID    *string `json:"tenant_id,omitempty"`
	Domain      string  `json:"domain"`
	Scenario    string  `json:"scenario"`
	Status      string  `json:"status"`
	DatasetPath string  `json:"dataset_path"`
	CreatedAt   string  `json:"created_at"`
	StartedAt   *string `json:"started_at,omitempty"`
	FinishedAt  *string `json:"finished_at,omitempty"`
	Error       *string `json:"error,omitempty"`
}

type GridMetrics struct {
	Precision float64 `json:"precision"`
	Recall    float64 `json:"recall"`
	F1        float64 `json:"f1"`
	PrAuc     float64 `json:"pr_auc"`
	Support   int     `json:"support"`
}

type GridSearchResult struct {
	ID        string                 `json:"id"`
	JobID     string                 `json:"job_id"`
	Algorithm string                 `json:"algorithm"`
	Params    map[string]interface{} `json:"params"`
	Metrics   GridMetrics            `json:"metrics"`
	IsOptimal bool                   `json:"is_optimal"`
	CreatedAt string                 `json:"created_at"`
}

func (s *OrchestrationService) CreateGridsearchJob(ctx context.Context, req GridSearchJobRequest) (*CreateJobResponse, error) {
	var resp CreateJobResponse
	if err := s.t.Do(ctx, "POST", "/v1/orchestration/gridsearch/jobs", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OrchestrationService) GetGridsearchJob(ctx context.Context, jobID string) (*GridSearchJob, error) {
	var resp GridSearchJob
	if err := s.t.Do(ctx, "GET", fmt.Sprintf("/v1/orchestration/gridsearch/jobs/%s", jobID), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OrchestrationService) ListGridsearchResults(ctx context.Context, jobID string) ([]GridSearchResult, error) {
	var resp []GridSearchResult
	if err := s.t.Do(ctx, "GET", fmt.Sprintf("/v1/orchestration/gridsearch/jobs/%s/results", jobID), nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

type PromoteRequest struct {
	TenantID string `json:"tenant_id"`
	Scenario string `json:"scenario"`
}

func (s *OrchestrationService) PromoteGridsearchJob(ctx context.Context, jobID string, req PromoteRequest) (*OrchestrationConfig, error) {
	var resp OrchestrationConfig
	if err := s.t.Do(ctx, "POST", fmt.Sprintf("/v1/orchestration/gridsearch/jobs/%s/promote", jobID), req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

type DatasetEntry struct {
	Domain string `json:"domain"`
	Path   string `json:"path"`
}

func (s *OrchestrationService) ListDatasets(ctx context.Context) ([]DatasetEntry, error) {
	var resp []DatasetEntry
	if err := s.t.Do(ctx, "GET", "/v1/orchestration/datasets", nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// =============================================================================
// V3 Integrated Pipeline (POST /v3/pipeline/run)
// =============================================================================

type PipelineRequest struct {
	Prompt               string   `json:"prompt"`
	ScenarioID           string   `json:"scenario_id,omitempty"`
	Algorithms           []string `json:"algorithms,omitempty"`
	Category             string   `json:"category,omitempty"`
	TargetProviderID     string   `json:"target_provider_id,omitempty"`
	EvolutionGenerations int      `json:"evolution_generations,omitempty"`
	SaberTrials          int      `json:"saber_trials,omitempty"`
}

type PipelineResponse struct {
	Unprotected    map[string]interface{} `json:"unprotected"`
	RedTeam        map[string]interface{} `json:"redteam"`
	Paladin        map[string]interface{} `json:"paladin"`
	OutputDefense  map[string]interface{} `json:"output_defense"`
	Evolution      map[string]interface{} `json:"evolution"`
	TotalLatencyMs uint64                 `json:"total_latency_ms"`
}

func (s *OrchestrationService) PipelineRun(ctx context.Context, req PipelineRequest) (*PipelineResponse, error) {
	var resp PipelineResponse
	if err := s.t.Do(ctx, "POST", "/v3/pipeline/run", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// =============================================================================
// V3 Military Orchestrator (POST /v3/military/orchestrate, GET /v3/military/status)
// =============================================================================

type MilitaryOrchestrateRequest struct {
	Text          string `json:"text"`
	ChannelLevel  uint8  `json:"channel_level,omitempty"`
	SourceDomain  string `json:"source_domain,omitempty"`
	TargetDomain  string `json:"target_domain,omitempty"`
	SessionID     string `json:"session_id,omitempty"`
}

type MilitaryModuleResult struct {
	Status    string          `json:"status"`
	RiskScore float64         `json:"risk_score"`
	LatencyMs uint64          `json:"latency_ms"`
	Summary   string          `json:"summary"`
	Details   interface{}     `json:"details"`
}

type MilitaryOrchestrateResponse struct {
	OverallRisk    float64                         `json:"overall_risk"`
	OverallStatus  string                          `json:"overall_status"`
	Modules        map[string]MilitaryModuleResult `json:"modules"`
	TotalLatencyMs uint64                          `json:"total_latency_ms"`
	Timestamp      string                          `json:"timestamp"`
}

type MilitaryModuleStatus struct {
	Name         string  `json:"name"`
	Available    bool    `json:"available"`
	LastAnalysis *string `json:"last_analysis,omitempty"`
}

type MilitaryStatusResponse struct {
	Modules       []MilitaryModuleStatus `json:"modules"`
	OverallHealth string                 `json:"overall_health"`
	Version       string                 `json:"version"`
}

func (s *OrchestrationService) MilitaryOrchestrate(ctx context.Context, req MilitaryOrchestrateRequest) (*MilitaryOrchestrateResponse, error) {
	var resp MilitaryOrchestrateResponse
	if err := s.t.Do(ctx, "POST", "/v3/military/orchestrate", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OrchestrationService) MilitaryStatus(ctx context.Context) (*MilitaryStatusResponse, error) {
	var resp MilitaryStatusResponse
	if err := s.t.Do(ctx, "GET", "/v3/military/status", nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

File: python/src/aegis/resources/orchestration.py
"""Orchestration resource — /v1/orchestration API group.

All `runs/*` endpoints always include v1/judge. v2/v3 detectors are layered
on top via in-process adapters (GuardNet / JBShield / CCFC / MULI / DPI / PII).
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from aegis.resources._base import AsyncResource, SyncResource

Scenario = Literal[
    "basic", "standard", "full", "advanced", "agent", "anomaly", "pii"
]
Mode = Literal["balanced", "recall", "precision", "strict", "bayes", "auto"]
EnsembleAlgorithm = Literal[
    "max",
    "arith_mean",
    "weighted_mean",
    "noisy_or",
    "majority_vote",
    "log_odds_sum",
    "strict_consensus",
    "two_stage_veto",
    "hierarchical_cascade",
]
AnomalyAlgorithm = Literal["zscore", "moving_average", "iqr", "isolation_forest"]


def _run_body(
    content: str,
    mode: Optional[Mode] = None,
    algorithm: Optional[EnsembleAlgorithm] = None,
    weights: Optional[Dict[str, float]] = None,
    threshold: Optional[float] = None,
    scenario: Optional[Scenario] = None,
) -> Dict[str, Any]:
    body: Dict[str, Any] = {"content": content}
    if mode is not None:
        body["mode"] = mode
    if algorithm is not None:
        body["algorithm"] = algorithm
    if weights is not None:
        body["weights"] = weights
    if threshold is not None:
        body["threshold"] = threshold
    if scenario is not None:
        body["scenario"] = scenario
    return body


class SyncOrchestration(SyncResource):
    """Synchronous orchestration resource."""

    # ---- runs ----
    def _run(self, path: str, body: Dict[str, Any]) -> Dict[str, Any]:
        resp = self._post(path, json=body)
        return resp.get("decision", resp)

    def run(self, content: str, scenario: Scenario = "full", **kwargs: Any) -> Dict[str, Any]:
        return self._run(
            "/v1/orchestration/runs",
            _run_body(content, scenario=scenario, **kwargs),
        )

    def basic(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/basic", _run_body(content, **kwargs))

    def standard(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/standard", _run_body(content, **kwargs))

    def full(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/full", _run_body(content, **kwargs))

    def advanced(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/advanced", _run_body(content, **kwargs))

    def agent(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/agent", _run_body(content, **kwargs))

    def anomaly(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/anomaly", _run_body(content, **kwargs))

    def pii(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return self._run("/v1/orchestration/runs/pii", _run_body(content, **kwargs))

    def anomaly_timeseries(
        self,
        value: float,
        history: List[float],
        algorithm: AnomalyAlgorithm = "zscore",
    ) -> Dict[str, Any]:
        return self._post(
            "/v1/orchestration/runs/anomaly/timeseries",
            json={"value": value, "history": history, "algorithm": algorithm},
        )

    # ---- configs ----
    def get_config(self, tenant_id: str, scenario: Scenario) -> Dict[str, Any]:
        return self._get(f"/v1/orchestration/configs/{tenant_id}/{scenario}")

    def upsert_config(
        self,
        tenant_id: str,
        scenario: Scenario,
        algorithm: EnsembleAlgorithm,
        mode: Mode,
        weights: Optional[Dict[str, float]] = None,
        thresholds: Optional[Dict[str, float]] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"algorithm": algorithm, "mode": mode}
        if weights is not None:
            body["weights"] = weights
        if thresholds is not None:
            body["thresholds"] = thresholds
        return self._put(
            f"/v1/orchestration/configs/{tenant_id}/{scenario}", json=body
        )

    # ---- gridsearch (synchronous smoke-test variant) ----
    def gridsearch(
        self,
        domain: str,
        scenario: Scenario,
        max_samples: Optional[int] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"domain": domain, "scenario": scenario}
        if max_samples is not None:
            body["max_samples"] = max_samples
        return self._post("/v1/orchestration/gridsearch", json=body)

    # ---- gridsearch jobs (async, DB-persisted) ----
    def create_gridsearch_job(
        self,
        domain: str,
        scenario: Scenario,
        max_samples: Optional[int] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"domain": domain, "scenario": scenario}
        if max_samples is not None:
            body["max_samples"] = max_samples
        return self._post("/v1/orchestration/gridsearch/jobs", json=body)

    def get_gridsearch_job(self, job_id: str) -> Dict[str, Any]:
        return self._get(f"/v1/orchestration/gridsearch/jobs/{job_id}")

    def list_gridsearch_results(self, job_id: str) -> List[Dict[str, Any]]:
        return self._get(f"/v1/orchestration/gridsearch/jobs/{job_id}/results")

    def promote_gridsearch_job(
        self, job_id: str, tenant_id: str, scenario: Scenario
    ) -> Dict[str, Any]:
        return self._post(
            f"/v1/orchestration/gridsearch/jobs/{job_id}/promote",
            json={"tenant_id": tenant_id, "scenario": scenario},
        )

    def list_datasets(self) -> List[Dict[str, Any]]:
        return self._get("/v1/orchestration/datasets")

    # ---- V3 Integrated Pipeline ----
    def pipeline_run(
        self,
        prompt: str,
        scenario_id: str = "custom",
        algorithms: Optional[List[str]] = None,
        category: str = "general",
        target_provider_id: Optional[str] = None,
        evolution_generations: int = 5,
        saber_trials: int = 50,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {
            "prompt": prompt,
            "scenario_id": scenario_id,
            "category": category,
            "evolution_generations": evolution_generations,
            "saber_trials": saber_trials,
        }
        if algorithms is not None:
            body["algorithms"] = algorithms
        if target_provider_id is not None:
            body["target_provider_id"] = target_provider_id
        return self._post("/v3/pipeline/run", json=body)

    # ---- V3 Military Orchestrator ----
    def military_orchestrate(
        self,
        text: str,
        channel_level: int = 2,
        source_domain: str = "MilitaryNet",
        target_domain: str = "PublicNet",
        session_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {
            "text": text,
            "channel_level": channel_level,
            "source_domain": source_domain,
            "target_domain": target_domain,
        }
        if session_id is not None:
            body["session_id"] = session_id
        return self._post("/v3/military/orchestrate", json=body)

    def military_status(self) -> Dict[str, Any]:
        return self._get("/v3/military/status")


class AsyncOrchestration(AsyncResource):
    """Asynchronous orchestration resource."""

    async def _run(self, path: str, body: Dict[str, Any]) -> Dict[str, Any]:
        resp = await self._post(path, json=body)
        return resp.get("decision", resp)

    async def run(self, content: str, scenario: Scenario = "full", **kwargs: Any) -> Dict[str, Any]:
        return await self._run(
            "/v1/orchestration/runs",
            _run_body(content, scenario=scenario, **kwargs),
        )

    async def basic(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/basic", _run_body(content, **kwargs))

    async def standard(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/standard", _run_body(content, **kwargs))

    async def full(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/full", _run_body(content, **kwargs))

    async def advanced(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/advanced", _run_body(content, **kwargs))

    async def agent(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/agent", _run_body(content, **kwargs))

    async def anomaly(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/anomaly", _run_body(content, **kwargs))

    async def pii(self, content: str, **kwargs: Any) -> Dict[str, Any]:
        return await self._run("/v1/orchestration/runs/pii", _run_body(content, **kwargs))

    async def anomaly_timeseries(
        self,
        value: float,
        history: List[float],
        algorithm: AnomalyAlgorithm = "zscore",
    ) -> Dict[str, Any]:
        return await self._post(
            "/v1/orchestration/runs/anomaly/timeseries",
            json={"value": value, "history": history, "algorithm": algorithm},
        )

    async def get_config(self, tenant_id: str, scenario: Scenario) -> Dict[str, Any]:
        return await self._get(f"/v1/orchestration/configs/{tenant_id}/{scenario}")

    async def upsert_config(
        self,
        tenant_id: str,
        scenario: Scenario,
        algorithm: EnsembleAlgorithm,
        mode: Mode,
        weights: Optional[Dict[str, float]] = None,
        thresholds: Optional[Dict[str, float]] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"algorithm": algorithm, "mode": mode}
        if weights is not None:
            body["weights"] = weights
        if thresholds is not None:
            body["thresholds"] = thresholds
        return await self._put(
            f"/v1/orchestration/configs/{tenant_id}/{scenario}", json=body
        )

    async def gridsearch(
        self,
        domain: str,
        scenario: Scenario,
        max_samples: Optional[int] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"domain": domain, "scenario": scenario}
        if max_samples is not None:
            body["max_samples"] = max_samples
        return await self._post("/v1/orchestration/gridsearch", json=body)

    async def create_gridsearch_job(
        self,
        domain: str,
        scenario: Scenario,
        max_samples: Optional[int] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"domain": domain, "scenario": scenario}
        if max_samples is not None:
            body["max_samples"] = max_samples
        return await self._post("/v1/orchestration/gridsearch/jobs", json=body)

    async def get_gridsearch_job(self, job_id: str) -> Dict[str, Any]:
        return await self._get(f"/v1/orchestration/gridsearch/jobs/{job_id}")

    async def list_gridsearch_results(self, job_id: str) -> List[Dict[str, Any]]:
        return await self._get(f"/v1/orchestration/gridsearch/jobs/{job_id}/results")

    async def promote_gridsearch_job(
        self, job_id: str, tenant_id: str, scenario: Scenario
    ) -> Dict[str, Any]:
        return await self._post(
            f"/v1/orchestration/gridsearch/jobs/{job_id}/promote",
            json={"tenant_id": tenant_id, "scenario": scenario},
        )

    async def list_datasets(self) -> List[Dict[str, Any]]:
        return await self._get("/v1/orchestration/datasets")

    async def pipeline_run(
        self,
        prompt: str,
        scenario_id: str = "custom",
        algorithms: Optional[List[str]] = None,
        category: str = "general",
        target_provider_id: Optional[str] = None,
        evolution_generations: int = 5,
        saber_trials: int = 50,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {
            "prompt": prompt,
            "scenario_id": scenario_id,
            "category": category,
            "evolution_generations": evolution_generations,
            "saber_trials": saber_trials,
        }
        if algorithms is not None:
            body["algorithms"] = algorithms
        if target_provider_id is not None:
            body["target_provider_id"] = target_provider_id
        return await self._post("/v3/pipeline/run", json=body)

    async def military_orchestrate(
        self,
        text: str,
        channel_level: int = 2,
        source_domain: str = "MilitaryNet",
        target_domain: str = "PublicNet",
        session_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {
            "text": text,
            "channel_level": channel_level,
            "source_domain": source_domain,
            "target_domain": target_domain,
        }
        if session_id is not None:
            body["session_id"] = session_id
        return await self._post("/v3/military/orchestrate", json=body)

    async def military_status(self) -> Dict[str, Any]:
        return await self._get("/v3/military/status")
